processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -149,4 +155,178 @@ public void remove()
         }
     );
   }
+
+
+
+  public static Filter convertToCNFFromQueryContext(Query query, Filter filter)
```

Comment:
Hmm, could this method take any appreciable time to run? 

If so it'd be more efficient to move this from the low level runners (where it is now) to the broker, probably run in the same places that run `optimize()`. It'd have to operate on DimFilters for that to work though.